src/xdp/trn_transit_xdp_hardware_offload.c

```diff
@@ -0,0 +1,548 @@
+int _version SEC("version") = 1;
+
+struct bpf_map_def SEC("maps") networks_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct network_key_t),
+	.value_size = sizeof(struct network_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(networks_offload_map, struct network_key_t, struct network_offload_t);
+
+struct bpf_map_def SEC("maps") vpc_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct vpc_key_t),
+	.value_size = sizeof(struct vpc_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(vpc_offload_map, struct vpc_key_t, struct vpc_offload_t);
+
+struct bpf_map_def SEC("maps") endpoints_offload_map = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct endpoint_key_t),
+	.value_size = sizeof(struct endpoint_offload_t),
+	.max_entries = 1000001,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(endpoints_offload_map, struct endpoint_key_t, struct endpoint_offload_t);
+
+struct bpf_map_def SEC("maps") interface_config_offload_map = {
+	.type = BPF_MAP_TYPE_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(struct tunnel_iface_t),
+	.max_entries = 1,
+	.map_flags = 0,
+};
+BPF_ANNOTATE_KV_PAIR(interface_config_offload_map, int, struct tunnel_iface_t);
+
+static __inline int trn_rewrite_remote_mac(struct transit_packet *pkt)
+{
+	/* The TTL must have been decremented before this step, Drop the
+	packet if TTL is zero */
+	if (!pkt->ip->ttl)
+		return XDP_DROP;
+
+	struct endpoint_offload_t *remote_ep;
+	struct endpoint_key_t epkey;
+	epkey.tunip[0] = 0;
+	epkey.tunip[1] = 0;
+	epkey.tunip[2] = pkt->ip->daddr;
+	/* Get the remote_mac address based on the value of the outer dest IP */
+	remote_ep = bpf_map_lookup_elem(&endpoints_offload_map, &epkey);
+
+	if (!remote_ep) {
+		return XDP_DROP;
+	}
+
+	trn_set_src_mac(pkt->data, pkt->eth->h_dest);
+	trn_set_dst_mac(pkt->data, remote_ep->mac);
+
+	if (pkt->ip->tos & IPTOS_MINCOST) {
+		return XDP_PASS;
+	}
+
+	return XDP_TX;
+}
+
+static __inline int trn_router_handle_pkt(struct transit_packet *pkt,
+					  __u32 inner_src_ip,
+					  __u32 inner_dst_ip)
+{
+	__be64 tunnel_id = trn_vni_to_tunnel_id(pkt->geneve->vni);
+	/* This is where we forward the packet to the transit router:  First lookup
+	the network of the inner_ip->daddr, if found hash and forward to the
+	transit switch of that network, OW forward to the transit router. */
+
+	struct network_key_t nkey;
+	struct network_offload_t *net;
+	nkey.prefixlen = 80;//需要修改
+	__builtin_memcpy(&nkey.nip[0], &tunnel_id, sizeof(tunnel_id));
+	nkey.nip[2] = inner_dst_ip % 65536;
+	net = bpf_map_lookup_elem(&networks_offload_map, &nkey);
+
+	if (net) {
+		//trn_update_ep_host_cache(pkt, tunnel_id, inner_src_ip);
+		pkt->rts_opt->rts_data.host.ip = pkt->ip->daddr;
+		__builtin_memcpy(pkt->rts_opt->rts_data.host.mac,
+				 pkt->eth->h_dest, 6 * sizeof(unsigned char));
+	}
+
+	if (net) {
+
+		if (net->nip[0] != nkey.nip[0] || net->nip[1] != nkey.nip[1]) {
```

Comment:
can this be a memcmp?

Reply:
We'd like to keep the original code here.